---
title: "Vignette ecophylo"
author: "Team bouclé"
date: "9 janvier 2021"
output:
  html_document: default
  pdf_document: default
  
---

<style>
body {
text-align: justify}
</style>

```{r setup, include=FALSE}
library(reticulate)
knitr::opts_chunk$set(echo = TRUE)
use_python(Sys.which("python3"))
```

## Rules for working in Rmarkdown with python.

*we will need to check the setup chunk for location of python3 binary across different os (ubuntu work with this code)*

### Python chunks calling R obj.

R chunks works just fine as usual.

```{r}
py_config()
a <- 42 # set a in R
```

This chunk only work when the markdown is knitted.
See that we can call R object when looking for them in 'r' namespace.

```{python}
print(r.a)
r.a += 1 # modify R variable in python
print(r.a)
b = 666 # set b in python
```

### R chunks calling Python obj.

See that we can call python object when looking for them in 'py' list.

```{r}
print(a) # call a in R after modification in python
print(py$b) # call b in R
# Also possible to call directly python function in r
rand <- import('numpy')$random # import python module numpy and submod random
rand$randint(3) # call of function randint
```

You can run python code in interactive mod in the console using the following 
function : *repl_python(quiet = T)*.
It set the console in 'python mode' so you can type in python commands. 
However you can't send lines from a chunk or a script.

To exit the consol in python, just type *exit* and tadaa the consol is back 
to R.

### Using Python modules in R.

```{r, eval=FALSE}
# Be carefull for idiotproof of ecophylo and int class !
eco <- import('ecophylo')
eco$timeframes(I=as.integer(3), T=2, a=0.3)
# make dataframes with pandas is same as R
pd <- import('pandas')
d <- list(col1 = c(1,2,3), col2 = c(4,5,6))
pd$DataFrame(d)
```

\newpage

# Serious Part #####

## Introduction

We introduce the Python package ecophylo dedicated to the coalescent-based simulation of neutral eco-evolutionary dynamics. Species assemblages and their shared ancestry can be simulated by jointly taking into account the influence of past demographic fluctuations and extinctions along with how divergent genotypes have introduced new species over time through speciation.

The model rests on two main components: (i) a demographic component driving stochastic changes in population sizes, structure and extinctions due to habitat availability possibly linked migration events, (ii) a mutation and protracted speciation component representing how divergent genotypes emerge and define new species over time. 

Here we show how to simulate communities and their phylogenetic relationships for a wide range of past demographic scenarios having affected whole assemblages of species. We also provide ways to produce many simulated datasets from prior distributions so as to allow Approximate Bayesian Computation (ABC) methods, to retrieve likely parameter values from the comparison of these simulations to observed diversity patterns. 


## Installation

*explain that ecophylo is a python package, that the current vignette shows us here how to use in r blablbla*

The packages can be installed from the public repository Pypi with the `pip` tool.

You can also download the *tar.gz* file from the repository and install it with pip. You need to use python3 and install the dependencies before. Note that it is a development stage.

**impossible at this moment as the repository is private**

```{bash, eval=FALSE}
python3 -m pip install ecophylo
# NOT RUN
# wget -L https://github.com/thegreatlizzyator/ecophylo/blob/master/dist/ecophylo-0.0.5.tar.gz
# python3 -m pip install ecophylo-0.0.5.tar.gz
```

### Installation on Windows (the painfull way)

> For François : First create a new project in rstudio with **version control** from the [github](https://github.com/thegreatlizzyator/ecophylo). 

You will need to install `{reticulate}` package and python dependencies.

```{r, eval=FALSE}
# NOT RUN
install.packages('reticulate') # answer yes
library(reticulate)
# python dependencies
conda_install('r-reticulate', c('msprime','ete3','pandas'))
```

<!--
### Repo github

### CRAN

-->
## Simulating phylogenies 

### Constant size model

Our model rests on the fundamental hypothesis that fluctuations in the relative species abundances in a given habitat are driven by neutral drift dynamics, depending on the size of the assemblage (Hubbell, 2001). Thus, the dynamics of the assemblage they form can be represented by coalescence, i.e. by tracing the shared co-ancestry of extant individuals backwards in time until a single common ancestor is found (Kingman 1982).

We start by simulating the shared co-ancestry of n sampled individuals observed at present time in each assemblage using a backward simulation of the coalescent tree. Variant individuals are then sprinkled over the simulated genealogy conditionally to its topology and branch lengths. 


*EXPLAIN PROTRACTEDNESS - The probability that a mutation occurred on a specific branch of the genealogy, leading to a new variant individual, follows a Poisson distribution with parameter µ·B where µ is the point mutation rate and B is the length of the branch -- change to explain protracted parameters* 

Let us first simulate a case in which there are no past fluctuations nor migration events linking populations. This implies assuming that Jm the metacommunity effective size has been at equilibrium from present backward, until all lineages coalesced into their Most Recent Common Ancestor (MRCA). 


```{r}
ecophylo <- import('ecophylo')

n <- as.integer(10) # the sample size
Jm <- as.integer(500) # the size of the metacommunity (sensu Hubbell, 2001)
mu <- as.numeric(0.03) # the point mutation rate

tree <- ecophylo$simulate(n, Jm, mu, seed = as.integer(42), verbose = FALSE)

```

We can then compute summary statistics on the resulting phylogeny 

```{r}
library(ape)
library(picante)
library(vegan)

abund <- ecophylo$getAbund(tree, n)
phylo <- read.tree(text = tree$write())

# taxonomic diversity metrics:
S <- specnumber(abund) # species richness
H <- diversity(abund) # shannon's diversity
simp <- diversity(abund, "simpson") # simpson's diversity

# phylogenetic diversity metrics:
PD <- pd(matrix(abund, nrow =1, dimnames = list("", phylo$tip.label)), phylo) # Faith's phylogenetic diversity
MPD <- mpd(matrix(abund, nrow =1, dimnames = list("", phylo$tip.label)), cophenetic(phylo)) # Mean pairwise distance
MNTD <- mntd(matrix(abund, nrow =1, dimnames = list("", phylo$tip.label)), cophenetic(phylo)) # Mean nearest taxon distance

```


### Simulating past community-wide demographic fluctuations

However, past environmental changes such as alternating periods of contraction and expansion of suitable environmental conditions, may have shaped the evolutionary and ecological diversity of extant organisms. Specifically, climatic fluctuations have made environmental conditions alternatively common or rare over time. For instance, taxa may have undergone restriction of their distribution to local refugia during habitat contraction, from which they could expand when suitable habitat became more common. Yet, in the constant size model, Jm can only be approximately considered as the harmonic mean of the per-generation Jm dynamics. Thus by allowing Jm  to vary at specific ages of the past we can relax this equilibrium assumption and account for the discordance of demographic histories between distinct meta-communities.


Let us assume that the Jm has fluctuated in the past. 

```{r}
n <- as.integer(10) # the sample size
Jm <- as.integer(500) # the size of the metacommunity (sensu Hubbell, 2001)
mu <- as.numeric(0.03) # the point mutation rate

pastsizes <- list(as.integer(1000))
changetime <- list(as.integer(400))

tree <- ecophylo$simulate(n, Jm, mu, 
                          past_sizes = pastsizes, 
                          changetime = changetime, 
                          seed = as.integer(42), verbose = TRUE)
```




### Simulating sub-populations linked by vicariance and migration events



## Intensive simulations along priors




```{r, eval=FALSE}
# NOT RUN
truc <- eco$dosimuls(nsim = 1, sample_size = 100, comprior = list(1000,10e9),
             muprior = list(1e-3) , verbose = TRUE, seed = 42)
```

```{python, eval=FALSE}
# NOT RUN
eco.dosimuls(nsim = 5, sample_size = 100, comprior = [1000,10e9], 
             muprior = [1e-3] , verbose = True, seed = 42)
```


<!--
## Adding Migrations
-->

