% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
  \usepackage{amssymb}
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={ecophylo: Simulating and assessing eco-evolutionary dynamics under past environmental changes in Python and R},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\usepackage{setspace}
\doublespacing
\usepackage{lineno}
\linenumbers
\ifluatex
  \usepackage{selnolig}  % disable illegal ligatures
\fi

\title{\emph{ecophylo}: Simulating and assessing eco-evolutionary
dynamics under past environmental changes in Python and R}
\author{}
\date{\vspace{-2.5em}}

\begin{document}
\maketitle

Elizabeth Barthelemy\textsuperscript{1}*, Maxime
Jaunatre\textsuperscript{1}, and François Munoz\textsuperscript{1}

\textsuperscript{1} Université Grenoble Alpes, LIPhy, 140 Rue de la
Physique, 38402 Saint-Martin-d'Hères, France

Running title: ecophylo: simulating eco-evolutionary dynamics

Research article

Word count: XXXXX

(*) Corresponding author,
\href{mailto:elizabeth.barthelemy@univ-grenoble-alpes.fr}{\nolinkurl{elizabeth.barthelemy@univ-grenoble-alpes.fr}}

\newpage

\hypertarget{abstract}{%
\section{Abstract}\label{abstract}}

\begin{itemize}
\tightlist
\item
  We introduce the Python package ecophylo dedicated to coalescent-based
  simulation of eco-evolutionary dynamics. Species assemblages and their
  shared ancestry can be simulated by jointly taking into account the
  influence of past demographic fluctuations and extinctions along with
  how divergent genotypes have introduced new species over time through
  speciation.
\item
  The shared co-ancestry of present individuals is simulated backward in
  time using coalescent theory. Speciation events are then sprinkled
  over the simulated genealogy conditionally to its topology and branch
  lengths. The phylogenetic relationships amongst individuals and their
  abundances are finally obtained by merging paraphyletic clades into
  single species. Coalescent reconstruction of the genealogy of
  individuals can be simulated to represent past demographic
  fluctuations due to varying habitat availability, or include multiple
  communities linked by migration events.
\item
  The package includes tools to simulate large numbers of datasets and
  associated summary statistics, so that Approximate Bayesian
  Computation methods can be used to estimate parameter values of these
  processes.. Diverse patterns of taxonomic and phylogenetic
  compositions can be generated. The first version of the package allows
  simulating neutral coalescent genealogies, and will incorporate
  further non-neutral eco-evolutionary scenarios in future. The package
  can be used to explore how past demographic fluctuations have affected
  species abundances and phylogenetic relationships, and to estimate the
  parameters of these processes based on observed patterns. We provide
  step by step examples in both Python and R languages.
\end{itemize}

Key-words: eco-evolutionary modelling; community phylogeny; coalescent;
demographic stochasticity, ecological drift, extinction-speciation
dynamics

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

Observed species distributions and biodiversity patterns are shaped by
current ecological processes but also reflect the influence of past
evolutionary and biogeographic dynamics (Svenning et al.~2015). For
instance, alternating periods of contraction and expansion of suitable
environmental conditions should affect both demographic and
speciation-extinction dynamics in a given habitat over time (Barthelemy
et al.~Frontiers 2021, ref ?). Hence, a fundamental goal of biogeography
is to better understand how changes in suitable environmental conditions
due to past climatic and geomorphologic history have shaped current
biodiversity patterns (Bennett 1990). Historical biogeography approaches
typically investigate species appearance, extinction and migration
events without considering the role of demography and dispersal events
in community assembly over time (e.g., Yu, Harris \& He, 2010).
Meanwhile, population genetics and phylogeographic studies emphasize the
joint role of migration, mutation and drift in driving patterns of
genetic diversity in space and time (Avise, 2009). Specifically,
comparative phylogeographic approaches aim to grasp congruent or
differing influence of past historical events led by environmental
fluctuations on several co-occurring taxa (Arbogast and Kenagy 2001,
Swenson 2019, Overcast et al.~2019, 2020). Concurrently, the neutral
theory of biogeography has underlined the role of migration, speciation
and drift in shaping patterns of taxonomic diversity in space and time
(Hubbell, 2001). Thus, comprehensive and integrative modelling
approaches are still needed to address biodiversity dynamics at multiple
spatial and temporal scales, specifically regarding how elementary
mechanisms driving species coexistence and spatial dynamics in an
ecological perspective have influenced over time evolutionary
trajectories of species assemblages. However, most studies investigating
the intertwined influence of ecological and evolutive forces on the
relative abundances of species and associated eco-evolutive modelling
approaches have assumed near-equilibrium dynamics in the past (Hubbell,
2001). The neutral theory of biodiversity itself (Hubbell 2001) defines
the fundamental biodiversity number θ as equal to 2·Jm·ν (with Jm the
metacommunity effective size and ν the speciation rate per lineage per
generation) which has been estimated for various biomes across the globe
(ref?). However, this number assumes that Jm has been at equilibrium
from present backward, until all lineages coalesced into their Most
Recent Common Ancestor (MRCA). However, Jm can only be approximately
considered as the harmonic mean of the per-generation Jm dynamics. Thus
by allowing Jm to vary at specific ages of the past we can relax this
equilibrium assumption and account for the discordance of demographic
histories between distinct metacommunities. Yet, in the case where past
environmental variations (especially climate) have caused the sizes of
communities to fluctuate over time, these should have affected
ecological drift dynamics. Thus, in the case where these fluctuations
occur rapidly compared to the expected time-to-equilibrium of
speciation, migration and drift dynamics (for instance with long-lived
organisms with slow population dynamics), we expect that current
biodiversity patterns should retain the signature of past environmental
fluctuations.

\emph{intérêt de rajouter de la vicariance, de la migration avec
plusieurs communautés locales ?}

Thus, past environmental fluctuations have influenced the evolutionary
trajectory of communities by affecting the local coexistence of species,
their migration opportunities, extinction risks and altering speciation
pathways in time. Here we propose a novel simulation-based approach in
which we consider how assembly dynamics in finite and heterogeneous
environments (ecological perspective) affect speciation and extinction
dynamics over a long-term (evolutionary perspective), depending on
environmental changes over time. We devise an eco-evolutionary model,
resting on the reconstruction of the past co-ancestry of extant
individuals, and subsequent grouping of said individuals into distinct
species under a model of point mutations and protracted speciation of
monophyletic genetic groups. Because, only the ancestry of extant
observed individuals is traced back, coalescent methods do not require
simulating lineages with no descendants in the present and are thereby
much faster than their forward-in-time alternatives (Munoz et al.~2018).
Thus, without assuming to which species each individual belongs, we can
simulate the genealogies of individuals in an assemblage which can have
undergone fluctuations in the size of Jm, been split into
sub-assemblages linked by migration and/or vicariance or any combination
of these processes with possible differences in their relative
importance (Figure 1) (Hudson 2002, Kelleher and Lohse 2020). We see
multiple advantages to this approach as it may allow for designing in
silico experiments that have been growing in ecology to address the
possible outcome of (meta)community models. Simulating ecological
communities and their phylogenetic relations according to different
scenarios can help establish a benchmark against which to infer the
signatures of community-wide past biogeographic processes from the
resulting patterns of taxonomic and phylogenetic diversity. Thus,
simulation-based approaches hold great potential for inference with the
increase in computation power and the emergence of likelihood-free
inference, such as Approximate Bayesian Computation (ABC). We expose
here the logic and advantages of the approach to examine how past
community-wide non-equilibrium dynamics have shaped patterns of
taxonomic and phylogenetic diversity. The initial version of the method
presented here is suited to the simulation of neutral eco-evolutionary
dynamics, but future versions will allow further deviation from
neutrality. We introduce the method with the ecophylo package in Python
language which can be called in R to analyse simulated patterns of
diversity. The package includes options to simulate large numbers of
datasets over broad ranges of parameters and scenarios of past
demographic events and fluctuations. These methods are destined to be
used alongside ABC methods so as to estimate parameters of past
demographic fluctuation from the observation of actual patterns of
diversity.

\hypertarget{simulation-algorithm-in-ecophylo.simulate}{%
\section{\texorpdfstring{Simulation algorithm in
\emph{ecophylo.simulate}}{Simulation algorithm in ecophylo.simulate}}\label{simulation-algorithm-in-ecophylo.simulate}}

\hypertarget{coalescent-based-simulation-of-genealogies}{%
\subsection{Coalescent-based simulation of
genealogies}\label{coalescent-based-simulation-of-genealogies}}

Our model rests on the fundamental hypothesis that fluctuations in the
relative species abundances in a given habitat are driven by neutral
drift dynamics, depending on the size of the assemblage (Hubbell, 2001).
Following the Wright-Fisher model, we assume that all individuals shrunk
to their haplotypes can reproduce freely within the assemblage, and with
equal fitness (neutral assumption). The dynamics of the assemblage they
form can be represented by coalescence, i.e.~by tracing the shared
co-ancestry of extant individuals backwards in time until a single
common ancestor is found (Kingman 1982). Generations in the model are
discrete and non-overlapping, and a single coalescence event can happen
at a given generation. We can thus simulate the shared co-ancestry of n
sampled individuals observed at present time in each assemblage using a
backward simulation of the coalescent tree. Coalescence is an event such
as two lineages at generation t share the same ancestor at generation
t-1, which defines a bifurcating node in the stochastic genealogy of
individuals. Only the ancestry of individuals observed at present is
traced back, so that coalescent methods do not require simulating
lineages with no extant descendants and are thereby much faster than
their forward-in-time alternatives (Munoz et al.~2018). When n
\textless\textless{} Jm(T), the distribution of coalescence times can be
approximated as an exponential law with parameter λ proportional to
1/(2·Jm(T)) \{Wakeley, 2009\}. The topology of the genealogy is thus
influenced at each predefined period by the corresponding value of
Jm(T). Subpops, migration etc\ldots{} Thus, without assuming to which
species each individual belongs, we start by simulating the genealogies
of individuals in an assemblage experiencing past demographic
fluctuations and/or linked by vicariance or migration events using the
ms coalescent simulator (Hudson 2002; Kelleher, 2020).

\hypertarget{phylogenetic-reconstruction-and-species-abundances}{%
\subsection{Phylogenetic reconstruction and species
abundances}\label{phylogenetic-reconstruction-and-species-abundances}}

We sprinkled speciation events over the simulated genealogies
conditionally to their topology and branch lengths. The probability that
a mutation occured on a specific branch of the genealogy, leading to a
new variant descendant, followed a Poisson distribution with parameter
µ·B where µ is the point mutation rate and B is the length of the
branch. The descendants stemming from this branch defined a new
haplotype clade. Since an extant species should be a monophyletic clade
and include haplotypes differing from other species, all paraphyletic
clades of haplotypes at present were merged to form a single species.
Therefore, monophyletic lineages with distinct genotypes and older than
two generations were considered a distinct species (Manceau et
al.~2016). We derived thusly the phylogenetic relationships among
individuals as well as the number of individuals descending from a
speciation event in the genealogy, which defined the species abundance
in the sample at present.

\hypertarget{multi-species-eco-evolutionary-dynamics}{%
\section{Multi-species eco-evolutionary
dynamics}\label{multi-species-eco-evolutionary-dynamics}}

Here we show how to simulate communities and their phylogenetic
relationships for a wide range of past demographic scenarios having
affected whole assemblages of species using the \emph{ecophylo} package
in Python language. We provide examples in R language making use of the
functions provided in the package. We also provide ways to produce many
simulated datasets from prior distributions so as to allow Approximate
Bayesian Computation (ABC) methods to retrieve likely parameter values
from the comparison of these simulations to observed diversity patterns.

\hypertarget{simulating-multi-species-past-demographic-fluctuations}{%
\subsection{Simulating multi-species past demographic
fluctuations}\label{simulating-multi-species-past-demographic-fluctuations}}

The \emph{ecophylo} package essentially articulates itself around the
\emph{simulate} function. This function implements the above-mentioned
simulation algorithm and allows users to simulate a phylogeny in
\emph{give format?} given the desired parameter combinations accounting
for the demographic history of Jm.

In the following example, an assemblage of species and their
phylogenetic relationships, is simulated assuming that Jm has fluctuated
in the past over 4 predefined periods. Users can input the desired time
periods (in generations before present) by specifying the
\emph{parameter name} parameter, yet the \emph{timeframes} function can
be used to determine the periods while accounting for loss of resolution
in more ancient periods (Boitard et.al 2016 Plos Genetics). Changes in
sizes can be instantaneous or gradual if users provide a growth rate for
each period of time.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Chunk with pastsizes variations }
\end{Highlighting}
\end{Shaded}

We can then compute summary statistics on the resulting phylogeny. The
\emph{getAbund} function allows us to retrieve the number of individuals
descending from a speciation event in the genealogy, thus defining the
species abundance in the sample at present.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# chunk with getabund + picante/ape/vegan for sumstats}
\end{Highlighting}
\end{Shaded}

These summary statistics can be used to compare different
eco-evolutionary scenarios having yielded different patterns of extant
community composition. \emph{give an example chunk with diff past sizes
or even with different speciation models}

\emph{example with different protractedness tau (expected = the longer
tau the least rare species? plot this? )}

\hypertarget{simulating-multi-species-population-structure-and-history}{%
\subsection{Simulating multi-species population structure and
history}\label{simulating-multi-species-population-structure-and-history}}

\emph{introduce why we're doing this etc}

\hypertarget{discussion}{%
\section{Discussion}\label{discussion}}

Non-equilibrium modelling entails increasing complexity in the
demographic history of Jm, furthermore, the speciation model itself can
add considerable complexity to the demographic model depending on the
constraints it imposes on the speciation process (like protraction or
the strict respect of monophyly). In these situations, θ has no
tractable analytical solutions for a given set of demographic
parameters, as it has been largely shown in population genetics. A now
classical way to overcome this limit is by using the coalescent theory
within an approximate Bayesian computation framework. This framework
spares computing the explicit likelihood of the model by approximating
it through the relative proximity between the true dataset and datasets
simulated given a coalescent model.

\hypertarget{conclusion}{%
\section{Conclusion}\label{conclusion}}

\hypertarget{data-accessibility}{%
\section{Data accessibility}\label{data-accessibility}}

The \emph{ecophylo} package is available on
\url{https://github.com/thegreatlizzyator/ecophylo} The \emph{ecophylo}
package can be installed by applying the following command in Python and
can then be called into R provided the prior installation of
\emph{reticulate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# INSTALL ECOPHYLO HERE LOOOOL}
\end{Highlighting}
\end{Shaded}

\hypertarget{authors-contributions-statement}{%
\section{Authors' contributions
statement}\label{authors-contributions-statement}}

EB conceived the study and built the basic architecture of the
\emph{ecophylo} package. EB and MJ did substantially work on adding
functionalities, testing and cleaning code. All the authors contributed
substantially to setting up the framework and to write the manuscript.

\hypertarget{references}{%
\section{References}\label{references}}

\end{document}
